src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java

```diff
@@ -0,0 +1,203 @@
+		posestack.scale(0.8F, 0.8F, 1.0F);
+		
+		RenderSystem.applyModelViewMatrix();
+		
+		//draws the grist icon.
+		this.drawIcon(100, 4, type.getIcon());
+		
+		posestack.popPose();
+		RenderSystem.applyModelViewMatrix();
+		
+		return pTimeSinceLastVisible - this.lastChanged >= DISPLAY_TIME ? Toast.Visibility.HIDE : Toast.Visibility.SHOW;
+		
+		
+	}
+	
+	//modified version of drawIcon() from MinestuckScreen.java
+	private void drawIcon(int x, int y, ResourceLocation icon)
+	{
+		if(icon == null || Minecraft.getInstance() == null)
+			return;
+		
+		RenderSystem.setShader(GameRenderer::getPositionTexShader);
+		RenderSystem.setShaderTexture(0, icon);
+		
+		float scale = (float) 1 / 16;
+		
+		int iconX = 16;
+		int iconY = 16;
+		int iconU = 0;
+		int iconV = 0;
+		
+		BufferBuilder render = Tesselator.getInstance().getBuilder();
+		render.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX);
+		render.vertex(x, y + iconY, 0D).uv((iconU) * scale, (iconV + iconY) * scale).endVertex();
+		render.vertex(x + iconX, y + iconY, 0D).uv((iconU + iconX) * scale, (iconV + iconY) * scale).endVertex();
+		render.vertex(x + iconX, y, 0D).uv((iconU + iconX) * scale, (iconV) * scale).endVertex();
+		render.vertex(x, y, 0D).uv((iconU) * scale, (iconV) * scale).endVertex();
+		Tesselator.getInstance().end();
+	}
+	
+	//adds pDifference to the toast's current grist value.
+	private void addGrist(long pDifference) {
+		
+		this.difference += pDifference;
+		this.changed = true;
+		
+	}
+	
+	//Updates the grist value of any existing toasts, and if there aren't any of the same type, it instantiates a new one. NEVER use addToast() directly when adding a grist toast, ALWAYS use this method.
+	public static void addOrUpdate(ToastComponent pToastGui, GristType pType, long pDifference, EnumSource pSource, boolean pIncrease) {
+		
+		//try to find an existing toast with the same properties as the one being added.
+		GristToast gristToast = pToastGui.getToast(GristToast.class, pType.getTranslationKey() + pSource + String.valueOf(pIncrease));
+		
+		//if none can be found, add a new toast. If an existing toast *can* be found, update its grist amount using pDifference.
+		if (gristToast == null) {
+			pToastGui.addToast(new GristToast(pType, pDifference, pSource, pIncrease));
+		} else {
+			gristToast.addGrist(pDifference);
+		}
+		
+	}
+	
+	public static void sendGristMessage(Player player, GristSet set, GristToast.EnumSource source, boolean increase)
+	{
+		if(player.isLocalPlayer() == true)
+		{
+			List<GristAmount> reqs = set.getAmounts();
+			for(GristAmount pairs : reqs)
+			{
+				//the pair has to be split into two new variables because Map.Entry is immutable.
+				GristType type = pairs.getType();
+				long difference = pairs.getAmount();
+				
+				//ALWAYS use addOrUpdate(), and not addToast, or else grist toasts won't leave a running tally of the amount.
+				if (difference >= 0l)
```

Comment:
Replace 01 with either 1 or 0.

Reply:
It was a lowercase L not a 1, but either way it is unnecessary. Fixed.